fdbserver/BlobManifest.actor.cpp

```diff
@@ -461,9 +454,126 @@ class BlobManifestDumper : public ReferenceCounted<BlobManifestDumper> {
 				TraceEvent("BlobManfiestDumpError").error(e).log();
 				dprint("Manifest dumping error {}\n", e.what());
-				wait(tr.onError(e));
 				wait(BlobManifestFileSplitter::reset(splitter));
+				// wait to avoid spinning
+				wait(delay(SERVER_KNOBS->BLOB_MANIFEST_RETRY_INTERVAL));
+			}
+		}
+	}
+
+	// Read ranges and append to splitter. Assume the ranges can be processed within single transaction
+	ACTOR static Future<Version> dumpRanges(Reference<BlobManifestDumper> self,
+	                                        Reference<BlobManifestFileSplitter> splitter,
+	                                        std::vector<KeyRangeRef> ranges) {
+		state Transaction tr(self->db_);
+		try {
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			for (auto range : ranges) {
+				try {
+					state PromiseStream<RangeResult> rows;
+					state Future<Void> stream = tr.getRangeStream(rows, range, GetRangeLimits(), Snapshot::True);
+					loop {
+						RangeResult result = waitNext(rows.getFuture());
+						splitter->append(result);
+					}
+				} catch (Error& e) {
+					if (e.code() == error_code_end_of_stream) {
+						continue; // done and move to next range
+					}
+					throw;
+				}
+			}
+			Version readVersion = wait(tr.getReadVersion());
+			Value versionEncoded = BinaryWriter::toValue(readVersion, Unversioned());
+			splitter->append(KeyValueRef(blobManifestVersionKey, versionEncoded));
+			return readVersion;
+		} catch (Error& e) {
+			state Error err = e;
+			// Cannnot simply retry on transaction errors. Need to cleanup files for partial dump
+			// so wait on error to avoid spinning read, and throw it to caller
+			wait(tr.onError(e));
+			throw err;
+		}
+	}
+
+	// Start a transcation to read range and append to splitter. Number of rows are limited by maxRowsPerTransaction.
+	// It returns the last key that has been read.
+	ACTOR static Future<Key> dumpRange(Reference<BlobManifestDumper> self,
+	                                   Reference<BlobManifestFileSplitter> splitter,
+	                                   KeySelector begin,
+	                                   KeySelector end,
+	                                   Version maxVersion,
+	                                   int maxRowsPerTransaction) {
+		state Transaction tr(self->db_);
+		state int count = 0;
+		loop {
+			tr.setOption(FDBTransactionOptions::ACCESS_SYSTEM_KEYS);
+			tr.setOption(FDBTransactionOptions::PRIORITY_SYSTEM_IMMEDIATE);
+			tr.setOption(FDBTransactionOptions::LOCK_AWARE);
+			try {
+				state GetRangeLimits limits(SERVER_KNOBS->BLOB_MANIFEST_RW_ROWS);
+				limits.minRows = 0;
+				loop {
+					RangeResult result = wait(tr.getRange(begin, end, limits, Snapshot::True));
+					for (auto& row : result) {
+						// Check bgf/bgm keys and skip if they are generated after maxVersion
+						if (row.key.startsWith(blobGranuleFileKeys.begin)) {
```

Comment:
to make this more generic and the code a bit cleaner (and a tiny bit more performant as you don't have to check every key prefix), it might be more desirable to pass a filter function to this actor where the filter function (keyvalue) -> bool that returns whether the row should be included or not, and each of these code blocks would become that filter function

so then you could call like dumpRange(bgfKeys, filterBGFKeysFunction) and dumpRange(bghKeys, filterBGHFunction)

Reply:
fixed. thanks!